node/src/components/consensus/highway_core/state.rs

```diff
@@ -359,46 +366,54 @@ impl<C: Context> State<C> {
     pub(crate) fn validate_vote(&self, swvote: &SignedWireVote<C>) -> Result<(), VoteError> {
         let wvote = &swvote.wire_vote;
         let creator = wvote.creator;
-        wvote.panorama.validate(self)?;
-        let mut justifications = wvote.panorama.iter_correct();
-        if !justifications.all(|vh| self.vote(vh).timestamp <= wvote.timestamp) {
+        let panorama = &wvote.panorama;
+        let timestamp = wvote.timestamp;
+        panorama.validate(self)?;
+        if panorama.iter_correct(self).any(|v| v.timestamp > timestamp) {
             return Err(VoteError::Timestamps);
         }
-        match wvote.panorama.get(creator) {
-            Observation::Faulty => {
-                error!("Vote from faulty validator should be rejected in `pre_validate_vote`.");
-                return Err(VoteError::FaultyCreator); // Should be unreachable.
-            }
-            Observation::None if wvote.seq_number == 0 => (),
-            Observation::None => return Err(VoteError::SequenceNumber),
-            Observation::Correct(hash) => {
-                let prev_vote = self.vote(hash);
-                // The sequence number must be one more than the previous vote's.
-                if wvote.seq_number != 1 + prev_vote.seq_number {
-                    return Err(VoteError::SequenceNumber);
-                }
-                // The round exponent must only change one step at a time, and not within a round.
-                if prev_vote.round_exp != wvote.round_exp {
-                    let max_re = prev_vote.round_exp.max(wvote.round_exp);
-                    if prev_vote.round_exp + 1 < max_re
-                        || wvote.round_exp + 1 < max_re
-                        || prev_vote.timestamp >> max_re == wvote.timestamp >> max_re
-                    {
-                        return Err(VoteError::RoundLength);
-                    }
+        if wvote.seq_number != panorama.next_seq_num(self, creator) {
+            return Err(VoteError::SequenceNumber);
+        }
+        let r_id = round_id(timestamp, wvote.round_exp);
+        let opt_prev_vote = panorama[creator].correct().map(|vh| self.vote(vh));
+        let prev_round_exp = opt_prev_vote.map_or(self.params.init_round_exp(), |v| v.round_exp);
+        // The round exponent must only change one step at a time.
+        if prev_round_exp + 1 < wvote.round_exp || wvote.round_exp + 1 < prev_round_exp {
```

Comment:
I just wanted to reply that that could cause an underflow… but of course, in theory (but absolutely not in practice) the current code could already cause an _overflow_…

Reply:
Ah, of course. Yeah, and underflows are probably a much bigger risk here than overflows (especially since you can only change the round exponent by 1 per round, and the rounds get longer and longer when you increase it). I'm for leaving this as is, then.